agent/structs/service_definition.go

```diff
@@ -28,12 +30,17 @@ func (s *ServiceDefinition) NodeService() *NodeService {
 	return ns
 }
 
-func (s *ServiceDefinition) CheckTypes() (checks CheckTypes) {
+func (s *ServiceDefinition) CheckTypes() (checks CheckTypes, err error) {
```

Comment:
`else` is redundant if you flip the condition. You can consider it nit picking but the way I view it is that you first clear all the preconditions one after another without branching and then at the end - when all the pre-conditions are met - you execute the body. This way you can read the code top to bottom. 

The other thing is that `s.Checks = append(s.Checks, &s.Check)` has a side effect which you wouldn't expect when calling this method since it returns an entirely new `[]*CheckType`. I'd remove that and move it where it belongs. 

This also looks too object oriented and the other question is why would we filter empty and invalid checks on query instead of preventing them to be there in the first place? `s.Checks` should only have valid, non-empty `*CheckType` instances IMO.

```go
func (s *ServiceDefinition) CheckTypes() (checks CheckTypes, err error) {
	s.Checks = append(s.Checks, &s.Check)
	for _, check := range s.Checks {
		if check.Empty() {
			continue
		}
		if !check.Valid() {
			return nil, fmt.Errorf("invalid check definition:%v", check.ValidateMsg())
		}
		checks = append(checks, check)
	}
	return
}
```